### ISSUE TITLE
Choice of Radii in NS MNO script

### ISSUE THREAD

**Issue by cfos3120** [neuraloperator/markov_neural_operator#3]
I understand these hardcoded values:

```
# DISSIPATIVE REGULARIZATION PARAMETERS
# below, the number before multiplication by S is the radius in the L2 norm of the function space
radius = 156.25 * S # radius of inner ball
```
and 
```
radii = (radius, (525 * S) + radius) # inner and outer radii, in L2 norm of function space
```

are taken in relation to the L2 norm of the function space (this is just the L2 norm of the ground truth solution set, correct?).

How were these exact values, **156.25** and **525** determined? Were there any calculations/approximations involved or were these arbitrarily set?

**Comment by mliuschi** [comment]
Thanks for your question! The inner and outer radii are problem-dependent, in general. We would like to choose an inner radius such that all functions in the training set lie within this radius. Otherwise we'd be encouraging the model to ignore some important dynamics. So the 156.25 comes from our specific dataset. 

The outer radius is a bit less sensitive. I believe in this case we did a quick search for it, but in general we noticed that for sufficiently large outer radii, the model remained dissipative beyond this radius in practice.

**Comment by cfos3120** [comment]
How is the 156.25 calculated. Was it the mean L2 norm of the vorticity dataset? Did you change it when training the model for Cartesian Velocity components instead?

**Comment by mliuschi** [comment]
In this case, we want an inner radius that is at least as large as the _maximum norm_ (L2 or another one) of the functions in the dataset. If we choose the mean, then the "attraction shell" may intersect with the training set, which would provide conflicting optimization objectives in the same region of space.

When changing to velocity or any other dataset, we would need to recalculate this maximum norm.

**Comment by cfos3120** [comment]
So, to confirm this, I have your dataset 2D_NS_Re500.npy. And I run this line:
```
l2_spaces = np.linalg.norm(data[:,100:,...], axis=(2, 3))
print("min,mean,max", np.min(l2_spaces), np.mean(l2_spaces), np.max(l2_spaces))
```
This produces the following:
`min,mean,max 215.412 298.532 475.853`

Is it correct to compute this over the space domain? Because these numbers seem off from 156.25*S (10,000) and maximum radius you use (156.25+525)*S (i.e. 43,600)

**Comment by mliuschi** [comment]
@cfos3120 Hmm, I'm not able to reproduce your numbers. Are you sure you're using the raw data from Zenodo? Loading in the 128x128 data from Zenodo gives me an array of shape `(100, 501, 128, 128)`. Computing the norm over the last two dimensions as you have done gives me a maximum of `19342.1582`. Using `156.25 * S = 20000`, where `S = 128` is the resolution. If you downsample the original data to 64x64 and use `S = 64` instead, we should also see that the inner radius is just outside the outer bound of the data.

**Comment by cfos3120** [comment]
Could it be that the Zenodo link is outdated. I am using the one you provide in the readme: [https://zenodo.org/records/7495555](https://zenodo.org/records/7495555). Additionally, the dataset with shape (100, 501, 128, 128) is for Reynolds number 5000, Reynolds number 500 only has a shape of (1000, 501, 64, 64).

I have freshly downloaded all three datasets, and I have run this script:

```
for dataset in ['2D_NS_Re5000','2D_NS_Re500','2D_NS_Re40']:
        data = np.load(f"markov_neural_operator/data/{dataset}.npy")
        print(f"{dataset} has shape {data.shape} and the following L2 Norm Statistics")
        
        l2_spaces = np.linalg.norm(data[:,100:,...], axis=(2, 3))
        print(f"    Min: {np.min(l2_spaces)}")
        print(f"    Max: {np.max(l2_spaces)}")
        print(f"    Mean: {np.mean(l2_spaces)}")
```
Here is the output:

```
2D_NS_Re5000 has shape (100, 501, 128, 128) and the following L2 Norm Statistics
    Min: 470.5329284667969
    Max: 1100.477294921875
    Mean: 681.5128173828125
2D_NS_Re500 has shape (1000, 501, 64, 64) and the following L2 Norm Statistics
    Min: 215.41166223917176
    Max: 475.85303886511525
    Mean: 298.532389582885
2D_NS_Re40 has shape (200, 501, 64, 64) and the following L2 Norm Statistics
    Min: 110.01207579051508
    Max: 270.7950466198188
    Mean: 137.02144011881424
```

**Comment by mliuschi** [comment]
Thanks for your patience @cfos3120! And thanks for sharing your results. After investigating, you are right that the Zenodo link contains a different version of the dataset. I think the Zenodo version has been normalized. In this case, I would recommend setting your own radii based on the max norms you found above.

I will add a note to the README based on our discussion here. Thanks for your help, and feel free to reach out with any more questions!
